cpp/include/raft/util/cuda_utils.cuh

```diff
@@ -508,44 +508,307 @@ HDI double myATanh(double x)
 /** @} */
 
 /**
- * @defgroup LambdaOps Lambda operations in reduction kernels
+ * @defgroup LambdaOps Legacy lambda operations, to be deprecated
  * @{
  */
-// IdxType mostly to be used for MainLambda in *Reduction kernels
 template <typename Type, typename IdxType = int>
 struct Nop {
-  HDI Type operator()(Type in, IdxType i = 0) { return in; }
+  [[deprecated("Nop is deprecated. Use identity_op instead.")]] HDI Type
+  operator()(Type in, IdxType i = 0) const
+  {
+    return in;
+  }
 };
 
 template <typename Type, typename IdxType = int>
 struct SqrtOp {
-  HDI Type operator()(Type in, IdxType i = 0) { return mySqrt(in); }
+  [[deprecated("SqrtOp is deprecated. Use sqrt_op instead.")]] HDI Type
+  operator()(Type in, IdxType i = 0) const
+  {
+    return mySqrt(in);
+  }
 };
 
 template <typename Type, typename IdxType = int>
 struct L0Op {
-  HDI Type operator()(Type in, IdxType i = 0) { return in != Type(0) ? Type(1) : Type(0); }
+  [[deprecated("L0Op is deprecated. Use nz_op instead.")]] HDI Type operator()(Type in,
+                                                                               IdxType i = 0) const
+  {
+    return in != Type(0) ? Type(1) : Type(0);
+  }
 };
 
 template <typename Type, typename IdxType = int>
 struct L1Op {
-  HDI Type operator()(Type in, IdxType i = 0) { return myAbs(in); }
+  [[deprecated("L1Op is deprecated. Use abs_op instead.")]] HDI Type operator()(Type in,
+                                                                                IdxType i = 0) const
+  {
+    return myAbs(in);
+  }
 };
 
 template <typename Type, typename IdxType = int>
 struct L2Op {
-  HDI Type operator()(Type in, IdxType i = 0) { return in * in; }
+  [[deprecated("L2Op is deprecated. Use sq_op instead.")]] HDI Type operator()(Type in,
+                                                                               IdxType i = 0) const
+  {
+    return in * in;
+  }
 };
 
-template <typename Type>
+template <typename InT, typename OutT = InT>
 struct Sum {
-  HDI Type operator()(Type a, Type b) { return a + b; }
+  [[deprecated("Sum is deprecated. Use add_op instead.")]] HDI OutT operator()(InT a, InT b) const
+  {
+    return a + b;
+  }
 };
 
 template <typename Type>
 struct Max {
-  HDI Type operator()(Type a, Type b) { return myMax(a, b); }
+  [[deprecated("Max is deprecated. Use max_op instead.")]] HDI Type operator()(Type a, Type b) const
+  {
+    if (b > a) { return b; }
+    return a;
+  }
+};
+/** @} */
+
+/**
+ * @defgroup Functors Commonly used functors.
+ * The optional index argument is mostly to be used for MainLambda in reduction kernels
+ * @{
+ */
+
+struct identity_op {
```

Comment:
The original operator functions were jammed into `cuda_utils` back before we had built out a formal API for primitives and now that we are revamping and once again centralizing these, I think we should consider moving them (and protecting the HDI w/ another macro that is safe to use w/ gcc). 

I'm thinking something like `raft/core/operators.hpp` might be a great spot to put them more front-and-center. I'd also like see the operators in the sparse pairwise distances API consolidated there as well.

Reply:
This would be the macro to replace `HDI`, btw, which would be safe to use in both CUDA and non-cuda environments: https://github.com/rapidsai/raft/blob/11c51059c401d344097b86d9b9ab4a6f464e9410/cpp/include/raft/core/detail/macros.hpp#L40.

Reply:
Ok, there's one circular dependency issue to solve, though: the functors need `cuda_utils.cuh`, and `cuda_utils.cuh` has helpers that use the functors (e.g `warpReduce` when `reduce_op` isn't passed).

My suggestion would be to move all the reduction helpers to another header, e.g `device_utils.cuh` which already has reduction helpers. If we want to keep the deprecated ones for a while, we can replace `add_op` with a custom lambda there.

Reply:
+1 for `raft/core/operators.hpp` would be a perfect spot for the new operators! Also, I'd add `constexpr` to all constructors/methods.

Reply:
I'll also replace `mySqrt` and similar functions in the functors then, since the new header won't depend on `cuda_utils.cuh`